src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/InstrumentRecorder.cs

```diff
@@ -0,0 +1,200 @@
+        /// and the meter <paramref name="meter"/>.
+        /// </summary>
+        /// <param name="meter">The meter that produced the instrument required for recording the published measurements.</param>
+        /// <param name="instrumentName">The name of the instrument.</param>
+        public InstrumentRecorder(Meter meter, string instrumentName)
+        {
+            if (meter is null)
+            {
+                throw new ArgumentNullException(nameof(meter));
+            }
+
+            if (instrumentName is null)
+            {
+                throw new ArgumentNullException(nameof(instrumentName));
+            }
+
+            _measurements = new List<Measurement<T>>();
+
+            _meterListener = new MeterListener();
+            _meterListener.InstrumentPublished = (instrument, listener) =>
+            {
+                if (object.ReferenceEquals(instrument.Meter, meter) && instrument.Name == instrumentName &&
+                    (instrument is ObservableInstrument<T> || instrument is Instrument<T>))
+                {
+                    if (Interlocked.CompareExchange(ref _instrument, instrument, null) is null)
+                    {
+                        _isObservableInstrument = instrument is ObservableInstrument<T>;
+                        listener.EnableMeasurementEvents(instrument, state: null);
+                    }
+                }
+            };
+
+            _meterListener.SetMeasurementEventCallback<T>(OnMeasurementRecorded);
+            _meterListener.Start();
+        }
+
+        /// <summary>
+        /// Gets the <see cref="Instrument"/> that is being recorded.
+        /// </summary>
+        public Instrument? Instrument
+        {
+            get => _instrument;
+            private set => _instrument = value;
+        }
+
+        private void OnMeasurementRecorded(Instrument instrument, T measurement, ReadOnlySpan<KeyValuePair<string, object?>> tags, object? state)
+        {
+            var m = new Measurement<T>(measurement, tags);
+
+            lock (_measurements)
+            {
+                _measurements.Add(m);
+            }
+        }
+
+        /// <summary>
+        /// Gets the measurements recorded by this <see cref="InstrumentRecorder{T}"/>.
+        /// </summary>
+        /// <param name="clear">If true, the previously recorded measurements will be cleared.</param>
+        /// <returns>The measurements recorded by this <see cref="InstrumentRecorder{T}"/>.</returns>
+        public IEnumerable<Measurement<T>> GetMeasurements(bool clear = false)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(InstrumentRecorder<T>));
+            }
+
+            if (_isObservableInstrument)
+            {
+                _meterListener.RecordObservableInstruments();
```

Comment:
Oh I missed this. At some point in the past I thought there was an explicit RecordObservableInstrument() API. Doing it this way feels weird to me because now GetMeasurements() is mutating the list every time it is called. I think the design would be better if we have a separate explicit API that records observable instruments. For example imagine a user wants to capture measurements at several points, then verify them:
```C#
//arrange
using recorder = new InstrumentRecorder<int>(observableInstrument);

//act
SetState1();
recorder.GetMeasurements();
SetState2();
recorder.GetMeasurements();
SetState3();
recorder.GetMeasurements();

//assert
//I want to confirm that all three cases produced a measurement.
//but I can't use GetMeasurements() here because it changes the count to 4.
Assert.Equals(3, recorder.GetMeasurements().Length);
```
Users could of course workaround it by saving a snapshot of the measurements earlier, but needing to treat the API differently for observable and non-observable instruments feels awkward. Having API called 'GetXXX' change the thing it is getting also seems like it wouldn't align with our naming rules.

Reply:
I'll keep whatever I have in this PR and then we can talk about this one to know if we really need a new API.